pynsot/commands/cmd_protocols.py

```diff
@@ -0,0 +1,472 @@
+
+    """
+
+
+# Add
+@cli.command()
+@click.option(
+    '-u',
+    '--auth_string',
+    metavar='AUTH_STRING',
+    default='',
+    help='The authentication string (such as MD5 sum).',
+)
+@click.option(
+    '-a',
+    '--attributes',
+    metavar='ATTRS',
+    help='A key/value pair attached to this Protocol (format: key=value).',
+    multiple=True,
+    callback=callbacks.transform_attributes,
+)
+@click.option(
+    '-c',
+    '--circuit',
+    metavar='CIRCUIT',
+    help='The circuit that this protocol is running over.',
+)
+@click.option(
+    '-D',
+    '--device',
+    metavar='DEVICE',
+    type=types.NATURAL_KEY,
+    help=(
+        'Unique ID of the Device to which this Protocol is '
+        'running on.'
+    ),
+    required=True,
+)
+@click.option(
+    '-e',
+    '--description',
+    metavar='DESCRIPTION',
+    type=str,
+    help='The description for this Protocol.',
+)
+@click.option(
+    '-i',
+    '--interface',
+    metavar='INTERFACE',
+    type=str,
+    help=(
+        'The Interface this Protocol is running on. Either interface'
+        'or circuit must be populated.'
+    ),
+)
+@click.option(
+    '-s',
+    '--site-id',
+    metavar='SITE_ID',
+    type=int,
+    help=(
+        'Unique ID of the Site this Protocol is under.'
+        'If not set, this will be inherited off of the device\'s site'
+    ),
+    callback=callbacks.process_site_id,
+)
+@click.option(
+    '-t',
+    '--type',
+    metavar='TYPE',
+    type=types.NATURAL_KEY,
+    help='The type of the protocol.',
+    required=True,
+)
+@click.pass_context
+def add(ctx, auth_string, attributes, circuit, device, description, interface, site_id, type):
+    """
+    Add a new Protocol.
+
+    You must provide a Device hostname or ID using the -D/--device option.
+
+    You must provide the type of the protocol (e.g. OSPF, BGP, etc.)
+
+    If you wish to add attributes, you may specify the -a/--attributes
+    option once for each key/value pair.
+
+    """
+    data = ctx.params
+
+    if interface is None and circuit is None:
+        raise click.UsageError('Must have interface "-i" / "--interface" or circuit "-c" / "--circuit" populated')
+
+    if description is None:
+        data.pop('description')
+
+    ctx.obj.add(data)
+
+
+# List
+@cli.group(invoke_without_command=True)
+@click.option(
+    '-a',
+    '--attributes',
+    metavar='ATTRS',
+    help='A key/value pair attached to this Protocol (format: key=value).',
+    multiple=True,
+)
+@click.option(
+    '-u',
+    '--auth_string',
+    metavar='AUTH_STRING',
+    default='',
+    help='The authentication string (such as MD5 sum).',
+)
+@click.option(
+    '-c',
+    '--circuit',
+    metavar='CIRCUIT',
+    help='The circuit that this protocol is running over.',
+)
+@click.option(
+    '-d',
+    '--delimited',
+    is_flag=True,
+    help='Display set query results separated by commas vs. newlines.',
+    default=False,
+    show_default=True,
+)
+@click.option(
+    '-e',
```

Comment:
For reference, `-e` for description makes sense here because it conflicts with `-d/--delimited` for `list`. Delimited output is only an option on resource objects that can have attributes. I hope that makes sense compared to what can seem like a conflicting difference with ProtocolType (which can't have attributes and so doesn't support a `-d/--delimited` flag).